File: tool/replay_ipv6.py
#!/usr/bin/env python3
"""Send an RTP-MIDI Note On over IPv6 link-local (like Apple does)."""

import socket
import struct
import time
import sys

# Cubase Mac's link-local IPv6
REMOTE = "fe80::14bd:9902:492a:f2a3"
SCOPE_ID = 14  # en0 interface index
CONTROL_PORT = 5004
DATA_PORT = 5005

SSRC = 0xAABBCCDD
TOKEN = 0x12345678

def build_in(token, ssrc, name):
    name_bytes = name.encode() + b'\x00'
    return struct.pack('>HHIII', 0xFFFF, 0x494E, 2, token, ssrc) + name_bytes

def build_ck0(ssrc, t1):
    return struct.pack('>HHIB3x', 0xFFFF, 0x434B, ssrc, 0) + \
           struct.pack('>Q', t1) + b'\x00' * 16

def build_ck2(ssrc, ck1_data, t3):
    pkt = bytearray(36)
    struct.pack_into('>HHI', pkt, 0, 0xFFFF, 0x434B, ssrc)
    pkt[8] = 2
    pkt[12:28] = ck1_data[12:28]  # copy t1 and t2
    t3_hi = (t3 >> 32) & 0xFFFFFFFF
    t3_lo = t3 & 0xFFFFFFFF
    struct.pack_into('>II', pkt, 28, t3_hi, t3_lo)
    return bytes(pkt)

def build_bye(ssrc):
    return struct.pack('>HHIII', 0xFFFF, 0x4259, 2, 0, ssrc)

def make_addr(ip, port, scope):
    return (ip, port, 0, scope)

def main():
    scope = int(sys.argv[1]) if len(sys.argv) > 1 else SCOPE_ID

    # Bind to port 5004/5005 (matching Apple's convention exactly).
    ctrl = socket.socket(socket.AF_INET6, socket.SOCK_DGRAM)
    ctrl.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
    ctrl.bind(('::', CONTROL_PORT))
    data = socket.socket(socket.AF_INET6, socket.SOCK_DGRAM)
    data.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
    data.bind(('::', DATA_PORT))

    print(f"Local ports: {ctrl.getsockname()[1]}/{data.getsockname()[1]}")

    # 1. Control IN
    print("1. Control IN...")
    ctrl.sendto(build_in(TOKEN, SSRC, "Python Test"), make_addr(REMOTE, CONTROL_PORT, scope))
    resp, _ = ctrl.recvfrom(1024)
    print(f"   << {resp.hex()}")
    assert resp[2:4] == b'OK', "Expected OK"
    print("   Got control OK")

    # 2. Data IN
    print("2. Data IN...")
    data.sendto(build_in(TOKEN, SSRC, "Python Test"), make_addr(REMOTE, DATA_PORT, scope))
    resp, _ = data.recvfrom(1024)
    print(f"   << {resp.hex()}")
    assert resp[2:4] == b'OK', "Expected OK"
    print("   Got data OK")

    # 3. CK0
    print("3. CK0...")
    t1 = int(time.time() * 10000)  # 100us ticks
    data.sendto(build_ck0(SSRC, t1), make_addr(REMOTE, DATA_PORT, scope))
    ck1, _ = data.recvfrom(1024)
    print(f"   << CK1 ({len(ck1)}b)")

    # 4. CK2
    print("4. CK2...")
    t3 = int(time.time() * 10000)
    data.sendto(build_ck2(SSRC, ck1, t3), make_addr(REMOTE, DATA_PORT, scope))

    print("Session established. Waiting 5s...")
    time.sleep(5)

    # 5. MIDI Note On
    ts = int(time.time() * 10000) & 0xFFFFFFFF
    midi = struct.pack('>BBHII', 0x80, 0x61, 0x1000, ts, SSRC)
    midi += bytes([0x03, 0x90, 0x3C, 0x64])
    print(f"5. Sending MIDI: {midi.hex()}")
    data.sendto(midi, make_addr(REMOTE, DATA_PORT, scope))

    print("Waiting 10s...")
    time.sleep(10)

    # 6. BYE
    print("6. BYE")
    ctrl.sendto(build_bye(SSRC), make_addr(REMOTE, CONTROL_PORT, scope))
    data.sendto(build_bye(SSRC), make_addr(REMOTE, DATA_PORT, scope))

    ctrl.close()
    data.close()
    print("Done.")

if __name__ == "__main__":
    main()

File: tool/replay_ipv6.dart
/// Same as replay_apple.dart but using IPv6 link-local (like Apple does).
///
/// Usage: dart run tool/replay_ipv6.dart
library;

import 'dart:async';
import 'dart:io';
import 'dart:typed_data';

void main(List<String> args) async {
  // Cubase Mac's link-local IPv6 (from tcpdump capture).
  // Use numeric scope ID for local en0 interface.
  final scopeId = args.isNotEmpty ? args[0] : '14';
  final targetAddr = 'fe80::14bd:9902:492a:f2a3%$scopeId';

  print('Trying to resolve: $targetAddr');

  final target = InternetAddress(targetAddr);
  print('Resolved: ${target.address}, type: ${target.type}');

  final controlPort = 5004;
  final dataPort = 5005;

  // Bind to our own link-local address (required for fe80:: routing).
  final localAddr = InternetAddress('fe80::bc:8458:36c0:7b32%$scopeId');
  final ctrlSock = await RawDatagramSocket.bind(localAddr, 0);
  final dataSock = await RawDatagramSocket.bind(localAddr, ctrlSock.port + 1);

  print('Local ports: control=${ctrlSock.port}, data=${dataSock.port}');

  final ctrlQueue = _PacketQueue(ctrlSock);
  final dataQueue = _PacketQueue(dataSock);

  const ssrc = 0xAABBCCDD;
  const token = 0x12345678;

  // --- Handshake ---
  print('1. Control IN...');
  ctrlSock.send(_buildIN(token, ssrc, 'Dart IPv6'), target, controlPort);

  final ctrlOk = await ctrlQueue.waitFor(
      'ctrl OK', (d) => d.length >= 4 && d[2] == 0x4F && d[3] == 0x4B);
  if (ctrlOk == null) {
    print('FAILED');
    exit(1);
  }

  print('2. Data IN...');
  dataSock.send(_buildIN(token, ssrc, 'Dart IPv6'), target, dataPort);

  final dataOk = await dataQueue.waitFor(
      'data OK', (d) => d.length >= 4 && d[2] == 0x4F && d[3] == 0x4B);
  if (dataOk == null) {
    print('FAILED');
    exit(1);
  }

  // --- Clock sync ---
  print('3. CK0...');
  final t1 = DateTime.now().microsecondsSinceEpoch ~/ 100;
  dataSock.send(_buildCK0(ssrc, t1), target, dataPort);

  final ck1 = await dataQueue.waitFor('CK1',
      (d) => d.length >= 36 && d[2] == 0x43 && d[3] == 0x4B && d[8] == 1);
  if (ck1 == null) {
    print('FAILED');
    exit(1);
  }

  print('4. CK2...');
  final t3 = DateTime.now().microsecondsSinceEpoch ~/ 100;
  dataSock.send(_buildCK2(ssrc, ck1, t3), target, dataPort);

  print('Session established. Waiting 5s...');
  await Future.delayed(const Duration(seconds: 5));

  // --- Send MIDI ---
  final ts = DateTime.now().microsecondsSinceEpoch ~/ 100 & 0xFFFFFFFF;
  final midi = Uint8List(16);
  final mv = ByteData.sublistView(midi);
  midi[0] = 0x80;
  midi[1] = 0x61;
  mv.setUint16(2, 0x1000);
  mv.setUint32(4, ts);
  mv.setUint32(8, ssrc);
  midi[12] = 0x03;
  midi[13] = 0x90;
  midi[14] = 0x3C;
  midi[15] = 0x64;

  print('5. Sending MIDI: ${_hex(midi)}');
  dataSock.send(midi, target, dataPort);

  print('Waiting 10s...');
  await Future.delayed(const Duration(seconds: 10));

  print('6. BYE');
  ctrlSock.send(_buildBYE(ssrc), target, controlPort);
  dataSock.send(_buildBYE(ssrc), target, dataPort);
  await Future.delayed(const Duration(milliseconds: 500));

  ctrlQueue.dispose();
  dataQueue.dispose();
  ctrlSock.close();
  dataSock.close();
  print('Done.');
}

class _PacketQueue {
  final RawDatagramSocket _socket;
  final _packets = <Uint8List>[];
  final _waiters = <Completer<Uint8List>>[];

  _PacketQueue(this._socket) {
    _socket.listen((event) {
      if (event == RawSocketEvent.read) {
        final dg = _socket.receive();
        if (dg != null) {
          final data = Uint8List.fromList(dg.data);
          print('  << recv (${data.length}b): ${_hex(data)}');
          if (_waiters.isNotEmpty) {
            _waiters.removeAt(0).complete(data);
          } else {
            _packets.add(data);
          }
        }
      }
    });
  }

  Future<Uint8List?> waitFor(
      String label, bool Function(Uint8List) matcher) async {
    for (var i = 0; i < _packets.length; i++) {
      if (matcher(_packets[i])) {
        final p = _packets.removeAt(i);
        print('  >> $label OK');
        return p;
      }
    }
    final c = Completer<Uint8List>();
    _waiters.add(c);
    try {
      final data = await c.future.timeout(const Duration(seconds: 5));
      if (matcher(data)) {
        print('  >> $label OK');
        return data;
      }
      print('  >> $label: unexpected packet');
      return null;
    } on TimeoutException {
      print('  >> $label: timeout');
      return null;
    }
  }

  void dispose() {
    for (final c in _waiters) {
      if (!c.isCompleted) c.complete(Uint8List(0));
    }
  }
}

Uint8List _buildIN(int token, int ssrc, String name) {
  final nb = name.codeUnits;
  final d = Uint8List(16 + nb.length + 1);
  final v = ByteData.sublistView(d);
  v.setUint16(0, 0xFFFF);
  v.setUint16(2, 0x494E);
  v.setUint32(4, 2);
  v.setUint32(8, token);
  v.setUint32(12, ssrc);
  d.setRange(16, 16 + nb.length, nb);
  return d;
}

Uint8List _buildBYE(int ssrc) {
  final d = Uint8List(16);
  final v = ByteData.sublistView(d);
  v.setUint16(0, 0xFFFF);
  v.setUint16(2, 0x4259);
  v.setUint32(4, 2);
  v.setUint32(12, ssrc);
  return d;
}

Uint8List _buildCK0(int ssrc, int t1) {
  final d = Uint8List(36);
  final v = ByteData.sublistView(d);
  v.setUint16(0, 0xFFFF);
  v.setUint16(2, 0x434B);
  v.setUint32(4, ssrc);
  v.setUint32(12, (t1 >> 32) & 0xFFFFFFFF);
  v.setUint32(16, t1 & 0xFFFFFFFF);
  return d;
}

Uint8List _buildCK2(int ssrc, Uint8List ck1, int t3) {
  final d = Uint8List(36);
  final v = ByteData.sublistView(d);
  v.setUint16(0, 0xFFFF);
  v.setUint16(2, 0x434B);
  v.setUint32(4, ssrc);
  d[8] = 2;
  d.setRange(12, 28, ck1.sublist(12, 28));
  v.setUint32(28, (t3 >> 32) & 0xFFFFFFFF);
  v.setUint32(32, t3 & 0xFFFFFFFF);
  return d;
}

String _hex(Uint8List data) =>
    data.map((b) => b.toRadixString(16).padLeft(2, '0')).join(' ');
